services/cd-service/pkg/repository/transformer.go

```diff
@@ -201,13 +200,77 @@ func GetRepositoryStateAndUpdateMetrics(repo Repository) {
 
 // A Transformer updates the files in the worktree
 type Transformer interface {
-	Transform(context.Context, *State) (commitMsg string, changes *TransformerResult, e error)
+	Transform(context.Context, *State, TransformerContext) (commitMsg string, e error)
 }
 
-type TransformerFunc func(context.Context, *State) (string, *TransformerResult, error)
+type TransformerContext interface {
+	Execute(Transformer) error
+	AddAppEnv(app string, env string, team string)
+	DeleteEnvFromApp(app string, env string)
+}
+
+func RunTransformer(ctx context.Context, t Transformer, s *State) (string, *TransformerResult, error) {
+	runner := transformerRunner{
+		Context: ctx,
+		State:   s,
+		Stack:   [][]string{nil},
+	}
+	if err := runner.Execute(t); err != nil {
+		return "", nil, err
+	}
+	return runner.Stack[0][0], &TransformerResult{
+		ChangedApps:     runner.ChangedApps,
+		DeletedRootApps: runner.DeletedRootApps,
+		Commits:         runner.Commits,
+	}, nil
+}
 
-func (t TransformerFunc) Transform(ctx context.Context, state *State) (string, *TransformerResult, error) {
-	return (t)(ctx, state)
+type transformerRunner struct {
+	Context         context.Context
+	State           *State
+	Stack           [][]string
```

Comment:
Why do we need a `[][]string` for the stack instead of just `[]string`?
if we really need it, please add a small comment explaining the dimensions, like
 `// outer array is for foo, inner array is for bar`

Reply:
Sure, I added an explanation. It's not the most obvious thing ever, but since it is basically invisible when you're writing a transformer, I think it's ok.